src/Datadog.Trace/DuckTyping/DuckType.Methods.cs

```diff
@@ -308,7 +315,35 @@ private static void CreateMethods(TypeBuilder proxyTypeBuilder, Type proxyType,
                             }
 
                             // If the target parameter type is public or if it's by ref we have to actually use the original target type.
-                            targetParamType = UseDirectAccessTo(targetParamType) ? targetParamType : typeof(object);
+                            targetParamType = UseDirectAccessTo(proxyTypeBuilder, targetParamType) ? targetParamType : typeof(object);
+                            il.WriteSafeTypeConversion(proxyParamType, targetParamType);
+
+                            targetMethodParametersTypes[idx] = targetParamType;
+                        }
+                        else
+                        {
+                            if (NeedsDuckChaining(proxyParamType, targetParamType))
+                            {
+                                // Load the argument and cast it as Duck type
+                                il.WriteLoadArgument(idx, false);
+                                if (UseDirectAccessTo(proxyTypeBuilder, proxyParamType) && proxyParamType.IsValueType)
+                                {
+                                    il.Emit(OpCodes.Box, proxyParamType);
+                                }
+
+                                // We call DuckType.CreateCache<>.Create()
```

Comment:
Really small nit, but I think it may help other reviewers:

```suggestion
                                // We call DuckType.CreateCache<>.Create(object instance)
```